src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,698 @@
+                }))
+            }
+        });
+
+        let server = Server::bind(&addr)
+            .serve(svc)
+            .with_graceful_shutdown(handle_sigint());
+
+        if let Err(e) = server.await {
+            println!("server error: {}", e);
+        }
+
+        match self.applier.stop().await {
+            Ok(_) => println!("stopped applier"),
+            Err(e) => println!("failed to stop applier {}", e),
+        }
+
+        Ok(())
+    }
+}
+
+async fn handle_request(
+    addr: SocketAddr,
+    req: Request<Body>,
+    applier: Arc<apply::Applier>,
+) -> Result<Response<Body>, hyper::Error> {
+    let http_version = req.version();
+    let method = req.method().clone();
+    let cloned_uri = req.uri().clone();
+    let path = cloned_uri.path();
+
+    let resp = match http_version {
+        Version::HTTP_11 => {
+            match method {
+                Method::POST => {
+                    let mut resp = Response::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)...
+                    match req
+                        .into_body()
+                        .try_fold(Vec::new(), |mut data, chunk| async move {
+                            data.extend_from_slice(&chunk);
+                            Ok(data)
+                        })
+                        .await
+                    {
+                        Ok(body) => {
+                            let mut success = false;
+                            let mut req = apply::Request::new();
+                            match path {
+                                "/echo" => match echo::parse_request(&body) {
+                                    Ok(bb) => {
+                                        req.echo_request = Some(bb);
+                                        success = true;
+                                    }
+                                    Err(e) => {
+                                        resp = Response::builder()
+                                            .status(StatusCode::INTERNAL_SERVER_ERROR)...
+                                    }
+                                },
+                                _ => {
+                                    println!("unknown path {}", path);
+                                    resp = Response::builder()
+                                        .status(StatusCode::INTERNAL_SERVER_ERROR)...
+                                }
+                            }
+                            if success {
+                                match applier.apply(req).await {
+                                    Ok(rs) => resp = Response::new(Body::from(rs)),
+                                    Err(e) => {
+                                        resp = Response::builder()
+                                            .status(StatusCode::INTERNAL_SERVER_ERROR)...
+                                    }
+                                }
+                            }
```

Comment:
This could be structured in a better way. Like we don't need to keep a separate variable for success.